File: datastructures/labs/lab13_GraphBFS/GraphBFS.java
import java.util.*;

// Class representing a directed graph using adjacency lists
class GraphBFS {
    int V; // Number of Vertices
    LinkedList<Integer>[] adj; // adjacency lists

    // Constructor
    GraphBFS(int V) {
        this.V = V;
        adj = new LinkedList[V];

        for (int i = 0; i < adj.length; i++)
            adj[i] = new LinkedList<Integer>();
    }

    // To add an edge to graph
    void addEdge(int v, int w) {
        adj[v].add(w); // Add w to the list of v.
    } // addEdge

    // prints BFS traversal from a given starting node
    public void BFS(int node) {

        // mark all the vertices as unvisited (false)
        boolean visited[] = new boolean[V];

        // create a queue for BFS
        LinkedList<Integer> queue = new LinkedList<Integer>();

        // print the current node and mark it visited
        queue.add(node);
        visited[node] = true;

        while (queue.size() != 0) {
            // dequeue a vertex from queue and print it
            node = queue.poll();
            System.out.print(node + " ");

            // if an adjacent node is hasnt been visited,
            // enqueue it and mark it visited
            for (int neighbor : adj[node]) {
                if (!visited[neighbor]) {
                    queue.add(neighbor);
                    visited[neighbor] = true;
                }
            }
        }
    } // BFS

    public static void main(String args[]) {
        GraphBFS g = new GraphBFS(6);

        g.addEdge(0, 1);
        g.addEdge(0, 2);
        g.addEdge(2, 3);
        g.addEdge(2, 4);
        g.addEdge(4, 5);
        g.addEdge(1, 3);
        g.addEdge(3, 5);

        System.out.println("Breadth First Traversal");

        g.BFS(0);
    } // main
}

File: datastructures/python/labs/lab13_GraphBFS/GraphBFS.py
from collections import defaultdict


class Graph:
    # constructor
    def __init__(self):
        # default dictionary to store graph
        self.graph = defaultdict(list)

    def add_edge(self, vertex1, vertex2):
        """Function to add an edge to the graph from vertex1 to vertex2"""
        self.graph[vertex1].append(vertex2)

    def bfs(self, node):
        """Prints BFS traversal of a graph given a starting node."""
        # create a set to store all visited vertices
        visited = set()

        # create a queue for BFS
        queue = []

        # enqueue the source node and mark the it as visited
        queue.append(node)
        visited.add(node)

        while queue:

            # dequeue a node from the queue and print it
            node = queue.pop(0)
            print(node, end=" ")

            # if an adjacent node hasn't been visited,
            # enqueue it and mark it visited
            for neighbor in self.graph[node]:
                if neighbor not in visited:
                    queue.append(neighbor)
                    visited.add(neighbor)


if __name__ == "__main__":
    graph = Graph()
    graph.add_edge(0, 1)
    graph.add_edge(0, 2)
    graph.add_edge(1, 3)
    graph.add_edge(2, 3)
    graph.add_edge(2, 4)
    graph.add_edge(3, 5)
    graph.add_edge(4, 5)

    print("Breadth First Traversal")
    graph.bfs(0)
